docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1474 @@
+
+### The objective function
+
+#### Current situation
+
+The objective or criterion function is the function being optimized. Currently, it is
+called `criterion` in estimagic.
+
+The `criterion` function maps a pytree of parameters into a criterion value.
+
+An important feature of estimagic is that the same criterion function can work for
+scalar, least-squares and likelihood optimizers. Moreover, a criterion function can
+return additional data that is stored in the log file (if logging is active). All of
+this is achieved by returning a dictionary instead of just a scalar float.
+
+The conventions for the return value of the criterion function are as follows:
+
+- For the simplest case, where only scalar optimizers are used, `criterion` returns a
+  float or a dictionary containing the entry "value" which is a float.
+- For likelihood functions, `criterion` returns a dictionary that contains at least the
+  entry `"contributions"` which can be any pytree that can be flattened into a numpy
+  array of floats.
+- For least-squares problems, `criterion` returns a dictionary that contains at least
+  the entry `"root_contributions"` which can be any pytree that can be flattened into a
+  numpy array of floats.
+- In any case the returned dictionary can have an arbitrary number of additional keys.
+  The corresponding information will be stored in the log database (if logging is used).
+
+A few simple examples of valid criterion functions are:
+
+```python
+def sphere(params: np.ndarray) -> float:
+    return params @ params
+
+
+def dict_sphere(params: dict) -> float:
+    return params["a"] ** 2 + params["b"] ** 2
+
+
+def least_squares_sphere(params: np.ndarray) -> dict[str, Any]:
+    out = {"root_contributions": params, "p_mean": params.mean(), "p_std": params.std()}
+    return out
+```
+
+**Things we want to keep**
+
+- The fact that `params` can be arbitrary pytrees makes estimagic flexible and popular.
+  We do not need to restrict this type in any way because flattening the pytree gives us
+  a very precise type no matter how complicated the tree was.
+- Using the same criterion function for scalar, likelihood and least-squares optimizers
+  makes it easy to try out and compare very different algorithms with minimal code
+  changes.
+- We do not need to restrict the type of additional arguments of the criterion function.
+- The simplest form of our criterion functions is also compatible with scipy.optimize
+
+**Problems**
+
+- Most users of estimagic find it hard to write criterion functions that return the
+  correct dictionary. Therefore, they don't use the logging feature and we often get
+  questions about specifying least-squares problems correctly.
+- Internally we can make almost no assumptions about the output of a criterion function,
+  making the code that processes the criterion output very complex and full of if
+  conditions.
+- The best typing information we could get for the output of the criterion function is
```

Comment:
Not sure this qualifies as a problem under the status quo (no static typing). Seems more about process / implementation, i.e., a problem with the code base conditioning on wanting to make the move.

Reply:
Agree. Was also redundant.